File: apps/Offline-AI-Buddy/android/app/src/main/java/com/americangroupllc/offlineaibuddy/llm/LlamaService.kt
package com.americangroupllc.offlineaibuddy.llm

import com.americangroupllc.offlineaibuddy.core.models.ChatMessage
import com.americangroupllc.offlineaibuddy.core.models.ChatSession
import com.americangroupllc.offlineaibuddy.core.models.Language
import com.americangroupllc.offlineaibuddy.core.models.ModelManifest
import com.americangroupllc.offlineaibuddy.core.domain.PromptTemplates
import com.americangroupllc.offlineaibuddy.core.domain.TranslateOrchestrator
import com.americangroupllc.offlineaibuddy.core.storage.ModelStore
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.delay
import javax.inject.Inject
import javax.inject.Singleton

/**
 * App-side wrapper around [LlamaJni]. Mirrors the iOS `LlamaService`
 * API. v1 falls back to a canned echo stream when the JNI library
 * isn't loaded (CI / fork builds without llama.cpp submodule).
 */
@Singleton
class LlamaService @Inject constructor(
    private val store: ModelStore,
    private val manifest: ModelManifest,
) {
    private val jni = LlamaJni()
    @Volatile var modelLoaded: Boolean = false; private set

    suspend fun warmupIfModelPresent(): Boolean {
        val file = store.urlFor("${manifest.name}.gguf")
        if (!file.exists()) return false
        modelLoaded = jni.load(file.absolutePath, manifest.contextSize)
        return modelLoaded
    }

    fun generate(
        kind: ChatSession.Kind,
        language: Language,
        isKidSafe: Boolean,
        history: List<ChatMessage>,
        userInput: String,
        translateSrc: Language? = null,
        translateDst: Language? = null,
    ): Flow<String> = flow {
        val prompt = if (kind == ChatSession.Kind.TRANSLATE && translateSrc != null && translateDst != null) {
            TranslateOrchestrator.prompt(translateSrc, translateDst, userInput)
        } else {
            val base = PromptTemplates.prompt(kind, language, isKidSafe)
            val user = base.render(
                mapOf(
                    "user" to userInput,
                    "date" to java.time.LocalDate.now().toString(),
                    "audience" to "general",
                )
            )
            PromptTemplates.Prompt(base.system, user)
        }
        if (jni.isLoaded) {
            val out = jni.generate(prompt.system, prompt.userTemplate, maxTokens = 512)
            for (chunk in out.chunked(8)) {
                emit(chunk)
                delay(10)
            }
        } else {
            // Stub fallback so the UI still streams in CI.
            val pieces = listOf("(stub) ", "you said: ", userInput)
            for (p in pieces) {
                emit(p)
                delay(10)
            }
        }
    }.flowOn(Dispatchers.Default)
}

File: apps/Offline-AI-Buddy/ios/OfflineAIBuddy/Services/LlamaService.swift
import Foundation
import BuddyAICore
import SwiftUI

/// App-side wrapper around `BuddyAICore.LlamaRunner`. Exposes published
/// state for SwiftUI + a stable `generate(...)` API.
@MainActor
final class LlamaService: ObservableObject {

    @Published private(set) var modelLoaded: Bool = false
    @Published private(set) var generating: Bool = false

    let runner: LlamaRunner
    let store: ModelStore
    let manifest: ModelManifest

    init(runner: LlamaRunner, store: ModelStore, manifest: ModelManifest) {
        self.runner = runner
        self.store = store
        self.manifest = manifest
    }

    /// Called on app launch — loads the model into the runner if it's
    /// already on disk. Otherwise the OnboardingFlow downloads it.
    func warmupIfModelPresent() async {
        let url = store.url(forModelNamed: "\(manifest.name).gguf")
        guard FileManager.default.fileExists(atPath: url.path) else { return }
        do {
            try await runner.load(modelURL: url)
            modelLoaded = true
        } catch {
            modelLoaded = false
        }
    }

    /// Build a prompt + stream tokens. The caller owns the message
    /// history list it passes in.
    func generate(
        kind: ChatSession.Kind,
        language: Language,
        isKidSafe: Bool,
        history: [ChatMessage],
        userInput: String,
        translateSrc: Language? = nil,
        translateDst: Language? = nil
    ) -> AsyncStream<Token> {
        let prompt: PromptTemplates.Prompt
        if kind == .translate, let src = translateSrc, let dst = translateDst {
            prompt = TranslateOrchestrator.prompt(src: src, dst: dst, text: userInput)
        } else {
            let base = PromptTemplates.prompt(kind: kind, language: language, isKidSafe: isKidSafe)
            let rendered = base.render([
                "user": userInput,
                "date": ISO8601DateFormatter().string(from: Date()),
                "audience": "general",
            ])
            prompt = PromptTemplates.Prompt(system: base.system, userTemplate: rendered)
        }

        let messagesWithUser = history + [ChatMessage(role: .user, text: prompt.userTemplate)]
        generating = true
        let stream = AsyncStream<Token> { continuation in
            Task { [weak self] in
                guard let self else { continuation.finish(); return }
                let inner = await self.runner.generate(
                    systemPrompt: prompt.system,
                    messages: messagesWithUser
                )
                for await t in inner { continuation.yield(t) }
                continuation.finish()
                await MainActor.run { self.generating = false }
            }
        }
        return stream
    }
}
